apm-agent-plugins/apm-spring-webmvc-plugin/src/test/java/co/elastic/apm/agent/spring/webmvc/ExceptionHandlerInstrumentationWithExceptionResolverTest.java

```diff
@@ -0,0 +1,56 @@
+/*-
+ * #%L
+ * Elastic APM Java agent
+ * %%
+ * Copyright (C) 2018 - 2019 Elastic and contributors
+ * %%
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ * #L%
+ */
+package co.elastic.apm.agent.spring.webmvc;
+
+import co.elastic.apm.agent.spring.webmvc.testapp.exception_resolver.ExceptionResolverController;
+import co.elastic.apm.agent.spring.webmvc.testapp.exception_resolver.ExceptionResolverRuntimeException;
+import co.elastic.apm.agent.spring.webmvc.testapp.exception_resolver.RestResponseStatusExceptionResolver;
+import org.junit.Ignore;
+import org.junit.Test;
+import org.springframework.mock.web.MockHttpServletResponse;
+import org.springframework.test.context.ContextConfiguration;
+import org.springframework.test.web.servlet.MvcResult;
+import org.springframework.test.web.servlet.ResultActions;
+
+import static org.junit.Assert.assertEquals;
+import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.get;
+
+@ContextConfiguration(classes = {
+    ExceptionResolverController.class,
+    RestResponseStatusExceptionResolver.class})
+public class ExceptionHandlerInstrumentationWithExceptionResolverTest extends AbstractExceptionHandlerInstrumentationTest {
+
+    @Test
+    @Ignore
```

Comment:
Is this failing now? My original question was on this one- why is it failing? Is it an Exception we miss? Or does it now pass?

Reply:
It's failed because we miss exception:
17:16:50.329 [main] DEBUG org.springframework.beans.factory.support.DefaultListableBeanFactory - Returning cached instance of singleton bean 'org.springframework.boot.test.mock.mockito.MockitoBeans'
17:16:50.330 [main] DEBUG org.springframework.test.context.support.AbstractDirtiesContextTestExecutionListener - After test method: context [DefaultTestContext@6c284af testClass = ExceptionHandlerInstrumentationWithExceptionResolverTest, testInstance = co.elastic.apm.agent.spring.webmvc.ExceptionHandlerInstrumentationWithExceptionResolverTest@63a28987, testMethod = testCallApiWithExceptionThrown@ExceptionHandlerInstrumentationWithExceptionResolverTest, testException = java.lang.AssertionError: expected:<1> but was:<0>, mergedContextConfiguration = [WebMergedContextConfiguration@5890e879 testClass = ExceptionHandlerInstrumentationWithExceptionResolverTest, locations = '{}', classes = '{class co.elastic.apm.agent.spring.webmvc.testapp.exception_resolver.ExceptionResolverController, class co.elastic.apm.agent.spring.webmvc.testapp.exception_resolver.RestResponseStatusExceptionResolver}', contextInitializerClasses = '[]', activeProfiles = '{}', propertySourceLocations = '{}', propertySourceProperties = '{}', contextCustomizers = set[org.springframework.test.context.web.socket.MockServerContainerContextCustomizer@1757cd72, org.springframework.boot.test.context.filter.ExcludeFilterContextCustomizer@687e99d8, org.springframework.boot.test.json.DuplicateJsonObjectContextCustomizerFactory$DuplicateJsonObjectContextCustomizer@c86b9e3, org.springframework.boot.test.mock.mockito.MockitoContextCustomizer@0, org.springframework.boot.test.autoconfigure.properties.PropertyMappingContextCustomizer@0, org.springframework.boot.test.autoconfigure.web.servlet.WebDriverContextCustomizerFactory$Customizer@37f1104d], resourceBasePath = 'src/main/webapp', contextLoader = 'org.springframework.test.context.web.WebDelegatingSmartContextLoader', parent = [null]], attributes = map['org.springframework.test.context.web.ServletTestExecutionListener.populatedRequestContextHolder' -> true, 'org.springframework.test.context.web.ServletTestExecutionListener.resetRequestContextHolder' -> true]], class annotated with @DirtiesContext [false] with mode [null], method annotated with @DirtiesContext [false] with mode [null].
17:16:50.330 [main] DEBUG org.springframework.test.context.web.ServletTestExecutionListener - Resetting RequestContextHolder for test context [DefaultTestContext@6c284af testClass = ExceptionHandlerInstrumentationWithExceptionResolverTest, testInstance = co.elastic.apm.agent.spring.webmvc.ExceptionHandlerInstrumentationWithExceptionResolverTest@63a28987, testMethod = testCallApiWithExceptionThrown@ExceptionHandlerInstrumentationWithExceptionResolverTest, testException = java.lang.AssertionError: expected:<1> but was:<0>, mergedContextConfiguration = [WebMergedContextConfiguration@5890e879 testClass = ExceptionHandlerInstrumentationWithExceptionResolverTest, locations = '{}', classes = '{class co.elastic.apm.agent.spring.webmvc.testapp.exception_resolver.ExceptionResolverController, class co.elastic.apm.agent.spring.webmvc.testapp.exception_resolver.RestResponseStatusExceptionResolver}', contextInitializerClasses = '[]', activeProfiles = '{}', propertySourceLocations = '{}', propertySourceProperties = '{}', contextCustomizers = set[org.springframework.test.context.web.socket.MockServerContainerContextCustomizer@1757cd72, org.springframework.boot.test.context.filter.ExcludeFilterContextCustomizer@687e99d8, org.springframework.boot.test.json.DuplicateJsonObjectContextCustomizerFactory$DuplicateJsonObjectContextCustomizer@c86b9e3, org.springframework.boot.test.mock.mockito.MockitoContextCustomizer@0, org.springframework.boot.test.autoconfigure.properties.PropertyMappingContextCustomizer@0, org.springframework.boot.test.autoconfigure.web.servlet.WebDriverContextCustomizerFactory$Customizer@37f1104d], resourceBasePath = 'src/main/webapp', contextLoader = 'org.springframework.test.context.web.WebDelegatingSmartContextLoader', parent = [null]], attributes = map['org.springframework.test.context.web.ServletTestExecutionListener.populatedRequestContextHolder' -> true, 'org.springframework.test.context.web.ServletTestExecutionListener.resetRequestContextHolder' -> true]].

java.lang.AssertionError: 
Expected :1
Actual   :0
 <Click to see difference>


	at org.junit.Assert.fail(Assert.java:88)
	at org.junit.Assert.failNotEquals(Assert.java:834)
	at org.junit.Assert.assertEquals(Assert.java:645)
	at org.junit.Assert.assertEquals(Assert.java:631)
	at co.elastic.apm.agent.spring.webmvc.AbstractExceptionHandlerInstrumentationTest.assertExceptionCapture(AbstractExceptionHandlerInstrumentationTest.java:95)
	at co.elastic.apm.agent.spring.webmvc.ExceptionHandlerInstrumentationWithExceptionResolverTest.testCallApiWithExceptionThrown(ExceptionHandlerInstrumentationWithExceptionResolverTest.java:51)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)

Reply:
See if this one can be solved if you add a `@Advice.OnMethodExit` to [the Spring MVC advice](https://github.com/elastic/apm-agent-java/blob/9515ec757700af2dd31060cade2f226cb9957c76/apm-agent-plugins/apm-spring-webmvc-plugin/src/main/java/co/elastic/apm/agent/spring/webmvc/SpringTransactionNameInstrumentation.java#L103)

Reply:
In any case, isn't this one handle the Exception and return quietly?

Reply:
@kananindzya let's bring this PR to a merge: 

We shouldn't add the `@Advice.OnMethodExit` to the Spring MVC advice as suggested above (because we may get two errors for unhandled Exceptions - one from there and one from the `Servlet#service` or `Servlet#doFilter`).

When I look at the `DispatcherServlet` code, it seems you are getting the expected result: the `processDispatchResult` method [calls the `processHandlerException` method](https://github.com/spring-projects/spring-framework/blob/fcbca7d569256ca2abc793fbcbe2827737950179/spring-webmvc/src/main/java/org/springframework/web/servlet/DispatcherServlet.java#L1110), which should trigger your `RestResponseStatusExceptionResolver` and also [set the `javax.servlet.error.exception` attribute with the Exception](https://github.com/spring-projects/spring-framework/blob/master//spring-webmvc/src/main/java/org/springframework/web/servlet/DispatcherServlet.java#L1322) if the view is not empty. Then, [it removes that same attribute](https://github.com/spring-projects/spring-framework/blob/master/spring-webmvc/src/main/java/org/springframework/web/servlet/DispatcherServlet.java#L1119). However, just before it clears them, it calls `render`, which is a method we already instrument, which is great! So, all you need to do is apply the same logic of looking for those request attributes to here: https://github.com/elastic/apm-agent-java/blob/9498bee6d90f429fdae8c0408ed6963134d68daf/apm-agent-plugins/apm-spring-webmvc-plugin/src/main/java/co/elastic/apm/agent/spring/webmvc/DispatcherServletRenderInstrumentation.java#L72 and then you should get this error. It will be tied to the `render` span and not the transaction, but I think this is good enough for us. @felixbarny - any reason we should insist this error will be related to the transaction and not the span? We can set our own attribute that won't be cleared and look for it in the ServletApiAdvice, but not sure we need to.

Reply:
If we want to record the exception even if a view has been resolved for it, we should rather go the initial route of instrumenting `processHandlerException` instead of relying on the attribute to still be present on `render`. It would also be misleading if the `render` span contained the exception.

Reply:
> instead of relying on the attribute to still be present on `render`.

Spring deliberately sets these attributes to be present in render, and then removes them, so I assume it is safe enough to rely on that.

> it would also be misleading if the `render` span contained the exception

I agree. However it will be just as misleading to have the exception captured with regard to `processHandlerException`, because no error occurred there either...

Can we agree on that: the render instrumentation will add a `co.elastic.apm.spring.exception` attribute to the request with that Exception and the ServletApiAdvice will look for that in addition to the others?

Reply:
> I agree. However it will be just as misleading to have the exception captured with regard to processHandlerException, because no error occurred there either...

There would not be a dedicated span for `processHandlerException`. In the instrumentation, we would just capture the exception on the active transaction.

It just seems easier and more understandable to directly capture the exception in `processHandlerException` as opposed to set an exception attribute in the render advice and to get that back in the generic Servlet advice.

On the other hand, just relying on the generic `javax.servlet.error.exception` attribute has the advantage that it would potentially also work for other Servlet-based frameworks. But capturing the exception both by instrumenting `processHandlerException ` and looking for the `javax.servlet.error.exception` attribute has the potential of capturing the same error twice. There also seem to be semantical differences between `javax.servlet.error.exception` and `org.springframework.web.servlet.DispatcherServlet.EXCEPTION`

Reply:
> It just seems easier and more understandable to directly capture the exception in `processHandlerException` as opposed to set an exception attribute in the render advice and to get that back in the generic Servlet advice.

If you look at the code, this can also cause for capturing the same exception twice (set as one or the other request attribute and not cleared).

I think the only way to guarantee capturing the Exception only once is setting our own attribute.

Reply:
If we break the loop after finding the first exception, there is no danger of double-capturing.

Reply:
> If we break the loop after finding the first exception

I don't get it. If you capture the Exception in `processHandlerException` it can still be set as one of the known attributes of the request, which means that at the end of `doFilter` or `service` we will capture it again